### PR TITLE
Cleanup fulfilled requests after approve

### PR DESCRIPTION
**Pull request by vera-liu**
Issue:
 - [https://github.com/airbnb/superset/issues/1500](url)

Scenario: 
Different entries in permission requests could have been fulfilled after one role approval, they should be automatically cleaned up after approval

Example:
 1. user creates permission request on one table
 2. user is granted Admin role
 3. user's previous permission requests should be cleaned up

needs-review @bkyryliuk

### CODE REVIEW

File: tests/access_tests.py [hunk 1]
```diff
@@ -236,10 +236,24 @@ def create_access_request(ds_type, ds_name, role_name):
...
 
+        # Case 5. Create new access request to table
+        # Grant user alpha access
+        # Check if gamma access request was deleted
+
+        # create request on energy usage
+        access_request5 = create_access_request('table', 'energy_usage', TEST_ROLE_NAME)
+        ds_5_id = access_request5.datasource_id
+        # grant Admin role
+        self.get_resp(GRANT_ROLE_REQUEST.format('table', ds_5_id, 'gamma', 'Admin'))
```
Comment (bkyryliuk) [inline tests/access_tests.py]: this test case doesn't cover the cleanup. 
please add 2 test cases:
1.  2 DARs, 2 users, datasource is added to the role for 1 user and that resolves it for another
2. 2DARs, 2 users, user 1 became alpha, user 2 DAR approved and user 1 DAR cleaned up

File: superset/views.py [hunk 2]
```diff
@@ -1283,6 +1283,17 @@ def request_access(self):
             datasource_names=", ".join([o.name for o in datasources]),
         )
 
+    def clean_fulfilled_requests(self):
+        session = db.session
+        for r in session.query(DAR).all():
+            datasource = SourceRegistry.get_datasource(
+                r.datasource_type, r.datasource_id, session)
+            user = sm.get_user_by_id(r.created_by_fk)
+            if sm._has_view_access(
```
Comment (bkyryliuk) [inline superset/views.py]: it won't work if user was granted alpha / admin or database or schema access

File: tests/access_tests.py [hunk 3]
```diff
@@ -236,10 +239,46 @@ def create_access_request(ds_type, ds_name, role_name):
             'datasource_access', druid_ds_2_perm)
         self.assertIn(perm_view, druid_role.permissions)
 
+        # Case 5. Create new access requests for gamma and gamma2
```
Comment (bkyryliuk) [inline tests/access_tests.py]: make those cases in separate function please

File: tests/access_tests.py [hunk 4]
```diff
@@ -236,10 +239,46 @@ def create_access_request(ds_type, ds_name, role_name):
...
+
+        access_request7 = create_access_request(
+            'table', 'birth_names', TEST_ROLE_NAME, 'gamma')
+        access_request8 = create_access_request(
+            'table', 'birth_names', TEST_ROLE_NAME, 'gamma2')
+        ds_6_id = access_request7.datasource_id
+        # gamma becomes alpha
+        alpha_role = sm.find_role('Alpha')
+        gamma_user = sm.find_user(username='gamma')
+        gamma_user.roles.append(alpha_role)
```
Comment (bkyryliuk) [inline tests/access_tests.py]: you need to commit here to in order to add an actual role to the user

File: tests/access_tests.py [hunk 5]
```diff
@@ -236,10 +239,46 @@ def create_access_request(ds_type, ds_name, role_name):
...
+        # request by gamma2 is fulfilled
+        self.client.get(EXTEND_ROLE_REQUEST.format(
+            'table', ds_6_id, 'gamma2', TEST_ROLE_NAME))
+        access_requests = self.get_access_requests('gamma', 'table', ds_6_id)
```
Comment (bkyryliuk) [inline tests/access_tests.py]: smth wrong here, as I do not understand why request disappears here

**Comment by vera-liu** [comment]
PTAL @bkyryliuk

File: superset/utils.py [hunk 6]
```diff
@@ -63,10 +63,13 @@ class SupersetTemplateException(SupersetException):
     pass
 
 
-def can_access(security_manager, permission_name, view_name):
+def can_access(sm, permission_name, view_name, current_user, user=None):
     """Protecting from has_access failing from missing perms/view"""
     try:
```
Comment (bkyryliuk) [inline superset/utils.py]: @vera-liu  - can we simplify it is do this:
```
def can_access(sm, permission_name, view_name, user):
  return sm.is_item_public(permission_name, view_name) or sm._has_view_access(user, permission_name, view_name)
```
?

File: superset/views.py [hunk 7]
```diff
@@ -82,7 +86,7 @@ def datasource_access_by_name(
             return True
 
         schema_perm = utils.get_schema_perm(database, schema)
-        if schema and utils.can_access(sm, 'schema_access', schema_perm):
+        if schema and utils.can_access(sm, 'schema_access', schema_perm, g.user, g.user):
```
Comment (bkyryliuk) [inline superset/views.py]: Line too long (89 > 79 characters)

File: tests/access_tests.py [hunk 8]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
             'datasource_access',
             updated_override_me.permissions[0].permission.name)
 
+    def test_cleanup(self):
```
Comment (bkyryliuk) [inline tests/access_tests.py]: s/test_cleanup/test_clean_fulfilled_requests

File: tests/access_tests.py [hunk 9]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
...
+        # Check if access request for gamma at energy_usage was deleted
+
+        # gamma2 and gamma request table_role on energy usage
+        access_request1 = create_access_request(
+            session, 'table', 'energy_usage', TEST_ROLE_1, 'gamma2')
+        ds_1_id = access_request1.datasource_id
+        access_request2 = create_access_request(
+            session, 'table', 'energy_usage', TEST_ROLE_1, 'gamma')
+        # gamma gets test_role1
+        self.get_resp(GRANT_ROLE_REQUEST.format(
```
Comment (bkyryliuk) [inline tests/access_tests.py]: add an assert for the response status

File: tests/access_tests.py [hunk 10]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
...
+            session, 'table', 'birth_names', TEST_ROLE_1, 'gamma')
+        access_request4 = create_access_request(
+            session, 'table', 'birth_names', TEST_ROLE_2, 'gamma2')
+        ds_3_id = access_request3.datasource_id
+        # gamma becomes alpha
+        alpha_role = sm.find_role('Alpha')
+        gamma_user = sm.find_user(username='gamma')
+        gamma_user.roles.append(alpha_role)
+        session.commit()
+        self.client.get(EXTEND_ROLE_REQUEST.format(
```
Comment (bkyryliuk) [inline tests/access_tests.py]: assert response code

File: tests/access_tests.py [hunk 11]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
...
+        access_request1 = create_access_request(
+            session, 'table', 'energy_usage', TEST_ROLE_1, 'gamma2')
+        ds_1_id = access_request1.datasource_id
+        access_request2 = create_access_request(
+            session, 'table', 'energy_usage', TEST_ROLE_1, 'gamma')
+        # gamma gets test_role1
+        self.get_resp(GRANT_ROLE_REQUEST.format(
+            'table', ds_1_id, 'gamma', TEST_ROLE_1))
+        # extend test_role1 with access on energy usage
+        self.client.get(EXTEND_ROLE_REQUEST.format(
```
Comment (bkyryliuk) [inline tests/access_tests.py]: ditto

File: tests/access_tests.py [hunk 12]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
             'datasource_access',
             updated_override_me.permissions[0].permission.name)
 
+    def test_cleanup(self):
+        session = db.session
+        TEST_ROLE_1 = 'test_role1'
+        TEST_ROLE_2 = 'test_role2'
+        sm.add_role(TEST_ROLE_1)
+        sm.add_role(TEST_ROLE_2)
```
Comment (bkyryliuk) [inline tests/access_tests.py]: could u move this code into setUpClass ?

File: tests/access_tests.py [hunk 13]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
...
+
+        # Case 1. Gamma and gamma2 requested test_role1 on energy_usage access
+        # Gamma already has role test_role1
+        # Extend test_role1 with energy_usage access for gamma2
+        # Check if access request for gamma at energy_usage was deleted
+
+        # gamma2 and gamma request table_role on energy usage
```
Comment (bkyryliuk) [inline tests/access_tests.py]: could u please make all cases - separate functions, they look pretty independent.
I need to refactor mine too.

File: tests/access_tests.py [hunk 14]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
...
+
+        access_request3 = create_access_request(
+            session, 'table', 'birth_names', TEST_ROLE_1, 'gamma')
+        access_request4 = create_access_request(
+            session, 'table', 'birth_names', TEST_ROLE_2, 'gamma2')
+        ds_3_id = access_request3.datasource_id
+        # gamma becomes alpha
+        alpha_role = sm.find_role('Alpha')
+        gamma_user = sm.find_user(username='gamma')
+        gamma_user.roles.append(alpha_role)
```
Comment (bkyryliuk) [inline tests/access_tests.py]: could u add:
```
access_requests = self.get_access_requests('gamma', 'table', ds_3_id)
self.assertTrue(access_requests)
```

to prove that the requests exited before.

File: tests/access_tests.py [hunk 15]
```diff
@@ -147,44 +175,132 @@ def test_override_role_permissions_drops_absent_perms(self):
...
+        access_requests = self.get_access_requests('gamma', 'table', ds_7_id)
+        self.assertFalse(access_requests)
+        gamma_user = sm.find_user(username='gamma')
+        gamma_user.roles.remove(sm.find_role(SCHEMA_ACCESS_ROLE))
+
+        ds = session.query(models.SqlaTable).filter_by(
+            table_name='wb_health_population').first()
+        ds.schema = None
+        gamma_user.roles.remove(sm.find_role('test_role1'))
+        session.delete(sm.find_role(TEST_ROLE_1))
```
Comment (bkyryliuk) [inline tests/access_tests.py]: move this to the tearDownClass

File: tests/base_tests.py [hunk 16]
```diff
@@ -65,6 +65,13 @@ def __init__(self, *args, **kwargs):
                 appbuilder.sm.find_role('Gamma'),
                 password='general')
 
+        gamma2 = appbuilder.sm.find_user('gamma2')
```
Comment (bkyryliuk) [inline tests/base_tests.py]: could u create function `find_or_create(user, user_group)` - looks like we have a lot of repetitive code here.
[optional]

**Comment by bkyryliuk** [comment]
looks pretty neat, left couple comments. @vera-liu ^^

File: superset/utils.py (hunk 6 above)
Comment (vera-liu) [inline superset/utils.py]: it wasn't working for public access when user is himself but hasn't logged in

File: tests/access_tests.py (hunk 9 above)
Comment (vera-liu) [inline tests/access_tests.py]: Since approve function returns a redirect, there isn't a status code

**Comment by bkyryliuk** [comment]
LGTM once tested on staging for different types of users.

**Pull request merged**